lib/CommandHandler.js

```diff
@@ -34,21 +35,43 @@ class CommandHandler {
         let commandName = args.splice(0, 1)[0].toLowerCase();
         let command = this.commands.get(commandName);
 
-        if(command) {
+        if(!command) {
+            addTemporaryReaction(message, "❌");
+            return logger.debug(`Command '${commandName}' was not found.`);
+        }
+
+        if(!message.channel.guild && command.guildOnly) {
+            return message.channel.createMessage("This command must be run in a server!");
+        }
+
+        if(!config.developerIDs.includes(message.author.id)) {
             let lastRun = this.cooldowns.get(commandName + ":" + message.channel.id);
-            
-            if(!lastRun || Date.now() - lastRun > command.cooldown) {
-                this.cooldowns.set(commandName + ":" + message.channel.id, Date.now());
-                command.run(this.client, message, args);
+            if(lastRun && Date.now() - lastRun < command.cooldown && !config.developerIDs.includes(message.author.id)) {
+                addTemporaryReaction(message, "⏱");
+                return logger.debug(`Command '${commandName}' exceeded cooldowns in channel '${message.channel.id}'`);
             }
-            else {
-                logger.debug(`Command '${commandName}' exceeded cooldowns in channel '${message.channel.id}'`);
+
+            for(let permission of command.permissions) {
+                if(!message.member.permission.has(permission)) {
+                    addTemporaryReaction(message, "👮");
+                    return logger.debug(`Member '${message.author.id}' lacks permission '${permission}' needed to run command '${commandName}'`);
+                }
             }
         }
-        else {
-            logger.debug(`Command '${commandName}' was not found.`);
-        }
+
+        this.cooldowns.set(commandName + ":" + message.channel.id, Date.now());
+        command.run(this.client, message, args);
+        addTemporaryReaction(message, "✅");
     }
 }
 
+function addTemporaryReaction(message, reaction) {
+    if(!config.reactToCommands) return;
+    message.addReaction(reaction).then(() => {
+        setTimeout(function _removeReaction() {
+            message.removeReaction(reaction).catch(() => {});
+        }, 7000);
+    }).catch(() => {});
```

Comment:
I think this would look a lot cleaner with an async/await syntax. You can promise-ify `setTimeout` to make it work with an await.